design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,140 @@
+
+```bash
+kubectl create -f deploy/examples/cosi/bucketclass.yaml
+kubectl create -f deploy/examples/cosi/bucketclaim.yaml
+kubectl create -f deploy/examples/cosi/bucketaccessclass.yaml
+kubectl create -f deploy/examples/cosi/bucketaccess.yaml
+```
+
+#### Consuming COSI Bucket
+
+The user needs to mount the secret as volume created by `BucketAccess` to the application pod. The user can use the secret to access the bucket by parsing the mounted file.
+
+```yaml
+spec:
+  containers:
+      volumeMounts:
+        - name: cosi-secrets
+          mountPath: /data/cosi
+  volumes:
+  - name: cosi-secrets
+    secret:
+      secretName: ba-secret
+```
+
+```bash
+# cat /data/cosi/bucket_info.json
+
+```json
+    {
+      apiVersion: "v1alpha1",
+      kind: "BucketInfo",
+      metadata: {
+          name: "ba-$uuid"
+      },
+      spec: {
+          bucketName: "ba-$uuid",
+          authenticationType: "KEY",
+          endpoint: "https://rook-ceph-my-store:443",
+          accessKeyID: "AKIAIOSFODNN7EXAMPLE",
+          accessSecretKey: "wJalrXUtnFEMI/K...",
+          region: "us-east-1",
+          protocols: [
+            "s3"
+          ]
+      }
+    }
+```
+
+#### Coexistence of COSI and lib-bucket-provisioner
+
+Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can coexist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The lib-bucket-provisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.
+
+#### Transition from libbucket provisioner to COSI
+
+This applied to OBC with reclaim policy is `Retain` otherwise the bucket will be deleted when OBC is deleted. So no point in migrating the OBC with `Delete` reclaim policy.
+
+- First the user need to create a **COSI Bucket resource** pointing to the backend bucket.
+- Then user can create BucketAccessClass and BucketAccess using the COSI Bucket CRD.
+- Now the update application's credentials with BucketAccess secret, for OBC it was combination of secret and config map with keys word like AccessKey, SecretKey, Bucket, BucketHost etc. Here details in as JSON format in the [secret](#Consuming-COSI-Bucket).
+- Finally the user need to delete the existing OBC.
+
+### Points to remember
+
+#### Ceph COSI Driver Requirements
+
+- CephObjectStore should be deployed and running in the cluster
+- The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions
+- The COSI controller should be deployed in the cluster
+- Rook can able to manage multiple ceph cosi drivers
```

Comment:
Rook will deploy one Ceph COSI driver per k8s cluster in the same manner as Ceph CSI